internal/handlers/readiness.go

```diff
@@ -17,11 +17,22 @@ func NewReadinessHandler() *ReadinessHandler {
 }
 
 func (a *ReadinessHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ok := func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusOK) }
+	a.runIfReady(http.HandlerFunc(ok), w, r)
+}
+
+func (a *ReadinessHandler) WithMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a.runIfReady(next, w, r)
+	})
+}
+
+func (a *ReadinessHandler) runIfReady(next http.Handler, w http.ResponseWriter, r *http.Request) {
 	if !a.isEnabled {
 		w.WriteHeader(http.StatusServiceUnavailable)
 		return
 	}
-	w.WriteHeader(http.StatusOK)
+	next.ServeHTTP(w, r)
 }
 
 func (a *ReadinessHandler) Enable() {
```

Comment:
Small question is do we want to keep the log in this function? If we do, is it a good idea to log which API is enabled since this middleware is now used by multiple handlers?

Looks good otherwise! I'll wait until @filanov reviews

Reply:
if i understand correctly it mean that all the apis are enabled and now the service ready to serve requests

Reply:
Correct. The same `Enable` call applies to all the handlers we wrap in the middleware.

Reply:
Gotcha, thanks for the explanation!